### DISCUSSION TITLE
Hodge laplacian vs Ripser/Gudhi/...

### DISCUSSION THREAD

**Discussion by 4ouRetta** [tsitsvero/hodgelaplacians#5] [Q&A]
Here's the issue I'm having:

Based on [Hodge laplacians](https://www.stat.uchicago.edu/~lekheng/work/psapm.pdf) (Table 3), I'm trying to find betti numbers per threshold and compare them to Gudhi/Ripser/..

I'm using random 2D point cloud data.  
I am using [Hodge laplacians Package](https://github.com/tsitsvero/hodgelaplacians) to calculate Laplacians from data point and RipsComplex from gudhi wich mentioned in the hodge laplacian repository that I linked before.

```
L=[]

for epsilon in tqdm.tqdm(np.linspace(0.5,4,100)): 
  #epsilon = Connect two points if distance is smaller than epsilon
  skeleton_1  = gd.RipsComplex(points = points, max_edge_length=epsilon )
  rips_simplex_tree_1 = skeleton_1.create_simplex_tree(max_dimension=2) 
  rips_skeleton_gudhi_1 = rips_simplex_tree_1.get_skeleton(1)
  hl_1 = HodgeLaplacians(rips_skeleton_gudhi_1, maxdimension=1, mode='gudhi')
  f1=hl_1.getHodgeLaplacian(0).toarray()
  b=sc.linalg.null_space(f1).T.shape[0]
  L.append(b)
```

and this part is ripser and gudhi part to calculate betti0 curve from persistent diagram:

```
dgms=ripser(points)["dgms"]
dgms0=dgms[0]
pdm0 = DiagramSelector(use=True, point_type='finite')(dgms0)
B0=BettiCurve(resolution=100,sample_range=[0.5,4])(pdm0)
```

they are compatible for Betti 0
as you can see here
[Betti 0 form hodge laplacian and Gudhi Betti Curve](https://i.stack.imgur.com/t4dZy.png)

For Betti 1 here is code parts:

Hodge laplacian method:

```
L=[]
for epsilon in tqdm.tqdm(np.linspace(0.5,4,100)): 
  #epsilon Connect two points if distance is smaller than epsilon
  skeleton_1  = gd.RipsComplex(points = points, max_edge_length=epsilon )
  rips_simplex_tree_1 = skeleton_1.create_simplex_tree(max_dimension=2) 
  rips_skeleton_gudhi_1 = rips_simplex_tree_1.get_skeleton(1)
  hl_1 = HodgeLaplacians(rips_skeleton_gudhi_1, maxdimension=2, mode='gudhi')
  f1=hl_1.getHodgeLaplacian(1).toarray()
  b=sc.linalg.null_space(f1).T.shape[0]
  L.append(b)
```

Ripser/Gudhi Part:

```
dgms=ripser(points)["dgms"]
dgms1=dgms[1]
pdm1 = DiagramSelector(use=True, point_type='finite')(dgms1)
B1=BettiCurve(resolution=100,sample_range=[0.5,4])(pdm1)
```

Here are the results for Betti 1, but they are incompatible as you can see here 
[Betti 1 form hodge laplacian and Gudhi Betti Curve](https://i.stack.imgur.com/E1x6e.png).

I discovered the problem is that the Hodge Laplacian method calculated all holes and the holes did not die as the filteration proceeds as we would expect with Vietoris Rips. Do you have any suggestions on what I might be missing here and how to fix this ? I want to make Hodge results like ripser ones.

  
Additionally, these are the papers in which I found the Hodge laplacian method.

[main paper](https://www.nature.com/articles/ncomms10138)

[nice review](https://www.nature.com/articles/ncomms10138)

[related powerpoint](https://liacs.leidenuniv.nl/~dunjkov/QAlg/Lecture8/qa_lecture_wk46.pdf)

**Comment by tsitsvero** [comment]
I'll try to take a look when I have more time to revive the project...
